djautotask/api_rest.py

```diff
@@ -0,0 +1,360 @@
+                    messages.append(
+                        '{}.'.format(error_message.rstrip('.'))
+                    )
+
+            messages = ' The error was: '.join(messages)
+
+            msg = '{} {}'.format(primary_error_msg, messages)
+
+        except json.decoder.JSONDecodeError:
+            # JSON decoding failed
+            msg = 'An error occurred: {} {}'.format(response.status_code,
+                                                    error)
+        except KeyError:
+            # 'code' or 'message' was not found in the error
+            msg = 'An error occurred: {} {}'.format(response.status_code,
+                                                    error)
+        return msg
+
+    def build_api_base_url(self):
+        self.api_base_url = '{0}v{1}/{2}/'.format(
+            self.server_url,
+            settings.AUTOTASK_CREDENTIALS['rest_api_version'],
+            self.API,
+        )
+
+    def get_headers(self):
+        headers = {'Content-Type': 'application/json'}
+
+        if self.username:
+            headers['UserName'] = self.username
+        if self.password:
+            headers['Secret'] = self.password
+        if self.integration_code:
+            headers['ApiIntegrationCode'] = self.integration_code
+
+        return headers
+
+    def build_filter(self, **kwargs):
+        for condition in kwargs['conditions']:
+            condition_arr = condition.split(",")
+            filter_obj = self._build_filter_obj(*condition_arr)
+            self.filter_array.append(filter_obj)
+
+    def _build_filter_obj(self, field, value, op='eq'):
+        filter_obj = {"op": op, "field": field, "value": value}
+        return filter_obj
+
+    def fetch_resource(self, endpoint_url, payload, full_next_url=False,
+                       retry_counter=None,
+                       *args, **kwargs):
+        """
+        Issue a GET request to the specified REST endpoint.
+
+        retry_counter is a dict in the form {'count': 0} that is passed in
+        to verify the number of attempts that were made.
+        """
+        @retry(stop_max_attempt_number=self.request_settings['max_attempts'],
+               wait_exponential_multiplier=RETRY_WAIT_EXPONENTIAL_MULTAPPLIER,
+               wait_exponential_max=RETRY_WAIT_EXPONENTIAL_MAX,
+               retry_on_exception=retry_if_api_error)
+        def _fetch_resource(endpoint_url, payload, full_next_url=False,
+                            retry_counter=None, *args, **kwargs):
+            if not retry_counter:
+                retry_counter = {'count': 0}
+            retry_counter['count'] += 1
+
+            try:
+                if full_next_url:
+                    endpoint = endpoint_url
+                else:
+                    endpoint = self._endpoint(self.POST_QUERY)
+
+                logger.debug('Making POST request to {} with {}'.format(
+                    endpoint, payload))
+                response = requests.post(
```

Comment:
Replying here because GitHub doesn't allow me to reply to your note about GET requests not working when querying with `lastActivityDate`. 

I think it would be preferred to keep this a GET request if possible. I just tested a similar request in Postman successfully with 
```
https://webservices2.autotask.net/atservicesrest/v1.0/contacts/query/?search={"filter": [{"op": "eq", "field": "IsActive", "value": "true"}, {"op": "gt", "field": "lastActivityDate", "value": "2019-11-26T18:12:18.567Z"}]}
```
I formatted the date using `isoformat` which is what django-connectwise does to format it's lastUpdated date before sending to the API. I think this is why you were getting the "No message. Conversion failed when converting date and/or time from character string." error. 

So I suggest changing back to a GET request and making sure that the date you use in your request is iso formatted.

Reply:
I was wrong. I see now that the date is already getting formatted with `isoformat`. But it appears that the API has the conversion error when passing a date string in this format: "2020-12-23T20:42:23.224615+00:00". 

I don't see the error if I use a string in this format: "2020-12-23T20:42:23.224615Z". So maybe this is the way to go. Give that a try and let me know if you are still seeing problems. 

To get that date formatted you should be able to use something like:
```
        if sync_job_qset.count() > 1 and not self.full:
            last_sync_job_time = sync_job_qset.last().start_time.strftime('%Y-%m-%d %H:%M:%S%fZ')
            self.api_conditions.append(
                "lastActivityDate,{0},gt".format(last_sync_job_time)
            )
```

Reply:
As discussed in Zoom call, I agree with Cameron on this point.